src/app/ReadClient.cpp

```diff
@@ -0,0 +1,239 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::ReportData),
+                 err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
+    VerifyOrExit(apExchangeContext == mpExchangeCtx, err = CHIP_ERROR_INCORRECT_STATE);
+    err = ProcessReportData(std::move(aPayload));
+
+exit:
+    ChipLogFunctError(err);
+    ClearExistingExchangeContext();
+    MoveToState(ClientState::Initialized);
+    return;
+}
+
+CHIP_ERROR ReadClient::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadClient::ProcessReportData(System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    ReportData::Parser report;
+
+    bool isEventListPresent  = false;
+    bool suppressResponse    = false;
+    bool moreChunkedMessages = false;
+
+    System::PacketBufferTLVReader reader;
+
+    reader.Init(std::move(aPayload));
+    reader.Next();
+
+    err = report.Init(reader);
+    SuccessOrExit(err);
+
+    err = report.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = report.GetSuppressResponse(&suppressResponse);
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    SuccessOrExit(err);
+
+    err = report.GetMoreChunkedMessages(&moreChunkedMessages);
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    SuccessOrExit(err);
+
+    {
+        EventList::Parser eventList;
+
+        err = report.GetEventDataList(&eventList);
+        if (CHIP_NO_ERROR == err)
+        {
+            isEventListPresent = true;
+        }
+        else if (CHIP_END_OF_TLV == err)
+        {
+            isEventListPresent = false;
+            err                = CHIP_NO_ERROR;
+        }
+        SuccessOrExit(err);
+
+        if (isEventListPresent && nullptr != mpDelegate && !moreChunkedMessages)
+        {
+            chip::TLV::TLVReader eventListReader;
+            eventList.GetReader(&eventListReader);
+            err = mpDelegate->EventStreamReceived(mpExchangeCtx, &eventListReader);
+            SuccessOrExit(err);
+        }
+    }
+
+    if (!suppressResponse)
+    {
+        // TODO: Add status report support and correspond handler in ReadHandler, particular for situation when there
+        // are multiple reports
+    }
+    if (nullptr != mpDelegate && !moreChunkedMessages)
+    {
+        err = mpDelegate->ReportProcessed(this);
+    }
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::OnResponseTimeout(Messaging::ExchangeContext * apExchangeContext)
+{
+    ChipLogProgress(DataManagement, "Time out! failed to receive report data from Exchange: %d",
+                    apExchangeContext->GetExchangeId());
+    ClearExistingExchangeContext();
+    MoveToState(ClientState::Initialized);
+    // TODO: Notify the SDK consumer response is timed out.
```

Comment:
Yes, this needs to happen.  This class is completely unusable as things stand.

Reply:
Done, thanks